data/transactions/verify/txn.go

```diff
@@ -229,43 +250,46 @@ func txnGroupBatchPrep(stxs []transactions.SignedTxn, contextHdr bookkeeping.Blo
 	return groupCtx, nil
 }
 
-// stxnCoreChecks runs signatures validity checks and enqueues signature into batchVerifier for verification.
-func stxnCoreChecks(s *transactions.SignedTxn, txnIdx int, groupCtx *GroupContext, batchVerifier *crypto.BatchVerifier) *TxGroupError {
-	numSigs := 0
-	hasSig := false
-	hasMsig := false
-	hasLogicSig := false
+// checkTxnSigTypeCounts checks the number of signature types and reports an error in case of a violation
+func checkTxnSigTypeCounts(s *transactions.SignedTxn) (isStateProofTxn bool, err *TxGroupError) {
```

Comment:
how about returning `numSigs int, sigOrTxnType sigOrTxnType` and let the caller to decide on the error?
where
```
type sigOrTxnType int
const regularSig sigOrTxnType = 1
const multiSig sigOrTxnType = 2
const logicSig sigOrTxnType = 3
const stateProofTxn sigOrTxnType = 4
```
then  the caller might have very the same error condition checks as now

Reply:
It is more complex and we don't have a use case where two callers are interested in reporting different errors.

Reply:
but in fact you are re-doing 64 bytes comparisons again for both sig and msig. Maybe this is not slower than int comparison on modern processors, not sure.
Btw, the second use
```
	isStateProofTxn, err := checkTxnSigTypeCounts(stx)
	if err != nil || isStateProofTxn {
		if err != nil {
			return 0, err
		}
		return 0, nil
	}
```
could be simplified to
```
	isStateProofTxn, err := checkTxnSigTypeCounts(stx)
	if err != nil || isStateProofTxn {
		return 0, err
	}
```

Reply:
This is a tricky one. Returning err (instead of nil), when err == nil, the caller is receiving err != nil. I could not understand why.

Reply:
> but in fact you are re-doing 64 bytes comparisons again for both sig and msig. Maybe this is not slower than int comparison on modern processors, not sure.

Convinced.